### PR TITLE
Added possibility to reuse client as a core part of Reactive client

### PR DESCRIPTION
**Pull request by bednar**
This PR solve [#392](https://github.com/influxdata/influxdb-java/issues/392) by adding possibility to reuse influxdb-java client as a core part of [influxdb-java-reactive](https://github.com/bonitoo-io/influxdb-java-reactive) - **Reactive Java client for InfluxDB**.

#### Summary of changes
The `InfluxDBImpl` constructor has parameter for customise `Retrofit.Builder`. That builder is use in **influxdb-java-reactive** client for create reactive version of `InfluxDBService`.

### CODE REVIEW

**Comment by majst01** [comment]
Hi @bednar 

You are solving 3 Issues with one PR, from my understanding only your first point is related to rxjava.

We only want to have PR´s which are focused on one single Issue.

If i am wrong can you please elaborate more details.

**Comment by bednar** [comment]
Hi @majst01,

Yes, you are right. 

During implementation the reactive client I found two issues with `InfluxDBResultMapper`:
- `InfluxDBResultMapper` is not able to handle results which a different time precision then milliseconds
- `InfluxDBResultMapper` is not able to map Integer to Instant

I will split the current PR to three pull requests:

1. The `InfluxDBResultMapper` is able to map `Integer` to `Instant`
2. The `InfluxDBResultMapper` is able to handle results which a different time precision
3. Added possibility to reuse client as a core part of Reactive client

It will be ok for you?

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 1]
```diff
@@ -146,7 +169,7 @@ public InfluxDBImpl(final String url, final String username, final String passwo
       break;
     }
 
-    this.retrofit = new Retrofit.Builder().baseUrl(url).client(
+    this.retrofit = retrofitBuilder.baseUrl(url).client(
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Just a heads up: you may have here `ConcurrentModificationException` as you are adding a `converterFactory` to a shared and non thread-safe `retrofitBuilder`. Worst case scenatio if you don't trigger a CME, you will end up having multiple `converterFactory` [added to the internal list in retrofitBuilder](https://github.com/square/retrofit/blob/master/retrofit/src/main/java/retrofit2/Retrofit.java#L515).

File: src/main/java/org/influxdb/impl/InfluxDBResultMapper.java [hunk 2]
```diff
@@ -152,7 +202,7 @@ void throwExceptionIfResultWithError(final QueryResult queryResult) {
     });
   }
 
-  void cacheMeasurementClass(final Class<?>... classVarAgrs) {
+  public void cacheMeasurementClass(final Class<?>... classVarAgrs) {
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBResultMapper.java]: Please revert the access modifier from this method on this class and from all methods from other classes to the minimum required. There is no need to make this method public.

File: src/main/java/org/influxdb/impl/InfluxDBResultMapper.java [hunk 3]
```diff
@@ -172,13 +222,22 @@ void cacheMeasurementClass(final Class<?>... classVarAgrs) {
     }
   }
 
-  String getMeasurementName(final Class<?> clazz) {
+  public String getMeasurementName(final Class<?> clazz) {
     return ((Measurement) clazz.getAnnotation(Measurement.class)).name();
   }
 
+  public <T> ConcurrentMap<String, Field> getColNameAndFieldMap(final Class<T> clazz) {
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBResultMapper.java]: There is no need to have this method: you are not mocking it and now you are exposing an internal structure publicly to the external code, allowing it to be changed. **Please, revert this change.**

File: src/main/java/org/influxdb/impl/InfluxDBResultMapper.java [hunk 4]
```diff
@@ -263,9 +324,11 @@ String getMeasurementName(final Class<?> clazz) {
       if (value instanceof String) {
         instant = Instant.from(ISO8601_FORMATTER.parse(String.valueOf(value)));
       } else if (value instanceof Long) {
-        instant = Instant.ofEpochMilli((Long) value);
+        instant = Instant.ofEpochMilli(toMillis((Long) value, precision));
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBResultMapper.java]: Could you please do a tinny improvement here by casting to primitives instead of Object wrappers? If you do, then please change the method signature from `toMillis(final Long value, ...` to `toMillis(final long value, ...` (just to avoid unboxing).

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java (hunk 1 above)
Comment (bednar) [reply src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Yes, you are right. It could cause same errors as in #445, so I will use same approach to avoid potential problems:

```java
Retrofit.Builder clonedRetrofitBuilder = retrofitBuilder.baseUrl(url).build().newBuilder();
this.retrofit = clonedRetrofitBuilder.client(clonedOkHttpBuilder.build())
            .addConverterFactory(converterFactory).build();
```

**Comment by bednar** [comment]
@fmachado I made the required changes. Can you check it?

**Comment by majst01** [comment]
Hi @bednar 

This PR still modifies around precision handling, and instantiation of a new influxdb service. I doubt these are all related to the usage in a reactive environment.

As long as we have mixed modifications in this PR we will not accept that.

**Comment by bednar** [comment]
Hi @majst01, thanks for response. 

Today, I will create separate PR for handling precision and leave this PR only for instantiation of a new InfluxDB service.

**Comment by bednar** [comment]
Hi @majst01, @fmachado 

I created pull request (#501) for the InfluxDBResultMapper precision issue and also changed description of current PR to reflect changes.

Thanks for your reviews

File: src/main/java/org/influxdb/InfluxDBException.java [hunk 5]
```diff
@@ -126,7 +126,7 @@ public boolean isRetryWorth() {
     }
   }
 
-  private static InfluxDBException buildExceptionFromErrorMessage(final String errorMessage) {
+  public static InfluxDBException buildExceptionFromErrorMessage(final String errorMessage) {
```
Comment (majst01) [inline src/main/java/org/influxdb/InfluxDBException.java]: Why public, if you want to test this like below, write the test in the matching package.

Comment (majst01) [reply src/main/java/org/influxdb/InfluxDBException.java]: Please add a changelog.md entry with pointers where the usage of influxdb-java in a reactive environment is shown.

Comment (bednar) [reply src/main/java/org/influxdb/InfluxDBException.java]: Because I want to reuse this method in influxdb-java-reactive to parse errors from HTTP header X-Influx-Error. If its is a problem I can use something like this:

```java
String errorBody = String.format("{\"error\":\"%s\"}", errorHeader);
```

Can I let this method public?

Comment (bednar) [reply src/main/java/org/influxdb/InfluxDBException.java]: The CHANGELOG.md was updated.

**Comment by bednar** [comment]
Hi @majst01,

I changed method InfluxDBException#buildExceptionFromErrorMessage to private.

Thanks for your review
